crates/sui-config/src/dynamic_transaction_signing_checks.rs

```diff
@@ -0,0 +1,273 @@
+    )]
+    InvalidReturnType(String),
+    #[error("Failed to execute Starlark program: {0}")]
+    ExecutionError(ErrorKind),
+    #[error("Failed to load Starlark program: {0}")]
+    LoadingError(ErrorKind),
+    #[error("Check failed -- transaction denied")]
+    CheckFailure,
+}
+
+#[derive(Debug, Clone)]
+pub struct DynamicCheckRunnerContext {
+    module: AstModule,
+    globals: Globals,
+    loaded_program: String,
+}
+
+const DIALECT: Dialect = Dialect {
+    enable_def: true,
+    enable_lambda: true,
+    enable_keyword_only_arguments: false,
+    enable_positional_only_arguments: false,
+    enable_types: DialectTypes::Disable,
+    // NB: set loader to false to prevent any external loading
+    enable_load: false,
+    enable_load_reexport: false,
+    // NB: Allow for top level statements to be used (e.g., top-level `for`, `if`, etc.)
+    enable_top_level_stmt: true,
+    enable_f_strings: false,
+    _non_exhaustive: (),
+};
+
+impl DynamicCheckRunnerContext {
+    /// Create a new `DynamicCheckRunnerContext` with the given Starlark program
+    /// `starlark_program` string. This will parse and validate the program is syntactically
+    /// correct and will set up shared (immutable) state that can be reused. This will not
+    /// run the program.
+    ///
+    /// The `starlark_program` string should be a valid Starlark program that returns a boolean
+    /// value when run -- `True` in the case that the transaction should be allowed, or `False` if
+    /// the transaction should be denied. Any other return value other than `True` (including
+    /// errors) should be considered a denial.
+    pub fn new(starlark_program: String) -> Result<Self, DynamicCheckRunnerError> {
+        // Adds global functions and variables to the dialect (e.g., True, False, Maps, Lists, etc.)
+        // The full spec of what exactly is added here can be found here:
+        // https://github.com/bazelbuild/starlark/blob/master/spec.md#built-in-constants-and-functions
+        let globals = Globals::standard();
+        warn!(
+            "Dynamic transaction checks are enabled. Make sure that you intend to be running \
+            dynamic checks on transactions."
+        );
+        let module = AstModule::parse(STAR_INPUT_FILE_NAME, starlark_program.clone(), &DIALECT)
+            .map_err(|e| DynamicCheckRunnerError::LoadingError(e.into_kind()))?;
+        Ok(Self {
+            module,
+            globals,
+            loaded_program: starlark_program,
+        })
+    }
+
+    /// Run the Starlark program in `self` with the given transaction data, signatures, input
+    /// object kinds, and receiving objects.
+    pub fn run_predicate(
+        &self,
+        tx_data: &TransactionData,
+        tx_signatures: &[GenericSignature],
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+    ) -> Result<(), DynamicCheckRunnerError> {
+        let tx_data_json = serde_json::to_value(tx_data)
```

Comment:
Do we have to worry about any user defined data here? Mostly thinking about TypeInput (note that none of these have been validated yet)

Reply:
We shouldn't -- this has all passed the transaction validity checks at this point so e.g., `TypeInput`s are bounded in their depth.